Shim from the old Predicate/BiPredicate to java.util versions

rmarsch opened the pull request:
https://private.hubteam.com/opengrok/search?q=%22net.jodah.failsafe%22+%26%26+%28failIf+%7C%7C+retryIf+%7C%7C+abortIf%29&defs=&refs=&path=java&hist=&type=&project=all

There's at least the `Meetings` project that directly created `BiPredicate` instances and fed them in, so we need the compile time there. 

I made the other shims because I believe the method signatures could create runtime issues with the heavy usage of lambda shorthand syntax for providing the `Predicate` and `BiPredicate` functions? I wasn't 100% confident in this assessment but it seemed better to be cautious as a first go.

Commit:
Shim from the old Predicate/BiPredicate to java.util versions

stevie400 commented on src/main/java/net/jodah/failsafe/FailurePolicy.java:
```diff
@@ -119,14 +119,24 @@ public S handleIf(Predicate<? extends Throwable> failurePredicate) {
-  public S failIf(Predicate<? extends Throwable> failurePredicate) {
-    return handleIf(failurePredicate);
+  public <T extends Throwable> S failIf(net.jodah.failsafe.function.Predicate<T> failurePredicate) {
+    return handleIf(new Predicate<T>() {
```
Throughout:
```java
return handleIf(failurePredicate::test);
```

stevie400 reviewed (commented):
Even with lambdas, we need them at runtime until everything using them as been rebuilt in order to avoid no such method exceptions at runtime because bytecode is references against a specific method (with return + arg types). The lambda case is source compatible but not binary compatible.

Are there any other similar situations we may need to consider to avoid breaking things?

Commit:
Simplify predicate shimming

- Make a helper class to do the conversion which removes the need for anonymous classes
- Need to make the old versions of the classes not able to be generated from a lambda expression to prevent ambiguity for the compiler

stevie400 commented on src/main/java/net/jodah/failsafe/function/BiPredicate.java:
```diff
@@ -19,4 +19,5 @@
 @Deprecated
 public interface BiPredicate<T, U> {
   boolean test(T t, U u);
+  void dontAllowLambdaShorthand();
```
This will change how things compile, which is what we want, but what does it do for binary compatibility? Will code that previously compiled against this using lambdas continue to run, or will we see exceptions?

stevie400 replied:
Actually, lookingat the opengrok our codebase I think this will only affect one project and we could easily avoid breakage by having that implement `dontAllowLambdaShorthand` before merging this PR

stevie400 commented on src/main/java/net/jodah/failsafe/function/Predicates.java:
```diff
@@ -0,0 +1,11 @@
+package net.jodah.failsafe.function;
+
+public class Predicates {
+  public static <T, U> java.util.function.BiPredicate<T, U> convertToJavaBiPredicate(BiPredicate<T, U> failsafeBiPredicate) {
```
It doesn't matter too much, but maybe add these as a default method called `toJavaUtil` on the interfaces themselves. This is what Guava does for `Optional`

Commit:
Finalize predicates migration setup after running test of runtime

- Ran a test of code compiled against 1.0.3 with lambdas in a runtime of this 2.3.1 to see that the bytecode is compatible even when we've altered the `Predicate` and `BiPredicate` classes to no longer accept lamdba shorthands in 2.3.1. The test worked successfully against this code so we should be good to go at runtime.

Commit:
Bring old execution model back to help migrate

- We have several uses of the old execution model that will fail without providing it or a shim. Trying to shim to the new execution model proved difficult because of the number of breaking changes to the interfaces.
- Restore the old execution model, shifting internal classes to having the suffix `Old` where necessary.
- Deprecated tags are set for everything brought over to support this change
- Needed to shift the varargs `with` method on `Failsafe` out of the way to support the 1.x `with` methods. Upgrades will need to be 2-phased first going to `withMigration` and then we can remove the old usage of `with` and then rename `withMigration` back to `with`

rmarsch commented:
I went with pulling the old execution model back in. Shimming the async execution model to the new one was getting really complicated to reason about due to the number of breaking changes in the upgrade, I was losing confidence I was going to maintain functionality. The number of LOC needed to maintain the old model is certainly gross here, but easier to keep track of and trust.
